### PR TITLE
[Draft] Polkadot Development Overview (build)

### PR DESCRIPTION
**Pull request by CrackTheCode016**
Creating a development "funnel" that clarifies the paths one can take in order to start developing on Polkadot. It is not meant to be a guide for everything, rather a concise series of pointers to necessary content oriented for builders.

This PR was created mainly because there has been a lack of clarity for some of those who are new to the ecosystem but wish to either simply tinker with Polkadot or dive into it at a more professional level.

Broad goals: 

- [x] Clarify the role of Polkadot versus parachains from a dev POV
- [x] Discuss the various paradigms and provide direction depending on the paradigm chosen
- [x] Clarify the differences and relationships between **key** development tools within the Polkadot ecosystem: 
  - Substrate, Cumulus, and FRAME clarity and the output of each component
  - Creating parachains versus solo chains
  - Direct pallet versus smart contract comparison
- [x] Links galore - provide links to *everything*
- [x] Graphics / diagrams to showcase the differences and outcomes of each choice

### CODE REVIEW

**Comment by kianenigma** [comment]
You will probably find this useful: 

https://forum.parity.io/t/polkadot-substrate-cumulus-host-runtime-all-in-one-diagram/1836

And

https://mermaid.live/edit#pako:eNplU9uOmzAQ_RXXUpSXQAmXSMvDSpRtJarQRDgoq0IeHHAaq2Ajx1abRfx7HdhuSDKSJWZ8zlwO4xYWvCTQhzmbTFrKqPRBO5VHUpOpP2VESYGraddNJjnL2aHif4ojFhIsk5wBbSe1_yVwcwQrpi8oG6IXQ5vsSxq9LHfAMJ4BCjNUX4ghZzplIXdX5IgT9uCI_f70EES8oiWV53xU46N6jE-SCLBRkguKqyti6KRPkGTbAMUgUUzSmgADfEuC-OvutoOLJT18HSRBtgVIj0tKgEihhK6uaaGqVaVOowEIK8ddDXy0Wq6y7efVfYZ3nua8l70X8nDolRzLMgyQBlkaNI1uAdW8KvlYwzToIWGwjh6C6-8o0wfou0cGUvvXzajU1b3vMPqxySKmZeYNEXhPq0GO1zDefQAM0zSeUXjrJ7duGuQMzmBNRI1pqVevvdznsF-6HPr6syQHrCqZ663sNBTr34rOrIC-FIrMoGpKLMkLxVqx-n-wwewn52MX-i38C33PMT3HffLm9sJZ2E-e687gGfqONTdty7F0wJ7bruN2M_jWJ7BMz-pt7tiOu7A8TSB69biIh7fSP5nuH4i756U

**Comment by kianenigma** [comment]
I would love to review this when final. The message here is crucial and we should adjust. This type of low level work can also use the opinion of a multitude of people, as it will be an important message. Who is --sort of-- the master mind behind the existing plan? 

On which page of the Polkadot website will this be hosted? 

Looking at the figure, one issue that I can spot is that you are using `Slot` instead of `Core`.

**Comment by CrackTheCode016** [comment]
> I would love to review this when final. The message here is crucial and we should adjust. This type of low level work can also use the opinion of a multitude of people, as it will be an important message. Who is --sort of-- the master mind behind the existing plan?

I believe it was one of the previous TechEd for the most current iteration, although I believe there were several peopel working on it over the years (@DrW3RK could clarify more probably). 

> On which page of the Polkadot website will this be hosted? 

For reference, this page currently lives here: https://wiki.polkadot.network/docs/build-guide. My hope is that it can serve as as good reference for any other pages which may benefit from defining these paths to development.

> I would love to review this when final. The message here is crucial and we should adjust. This type of low level work can also use the opinion of a multitude of people

@kianenigma Once it is final, I'll request a review :) I agree, it is something that needs a lot more input and thought from everyone! And thank you for those resources, will implement them accordingly.

**Comment by CrackTheCode016** [comment]
Added some more content, removed repetitive stuff, and added the notion of four tracks: 

1. [Building Parachains]
2. [Building a Pallet] (incomplete)
3. [Developing Smart Contracts]
4. [Developing a dApp] (incomplete)

Generally, I think a developer who (assuming knows nothing) could now more easily discern where they should go for learning depending on their goals.

Some more graphics would be nice, as well as making sure it flows well.

File: docs/build/build-guide.md [hunk 1]
```diff
@@ -7,17 +7,51 @@ keywords: [build, how to build, develop, parachain, smart contract]
 slug: ../build-guide
 ---
 
+import Tabs from "@theme/Tabs";
+
+import TabItem from "@theme/TabItem";
+
 Polkadot is a blockchain protocol with two goals: providing **shared security** among all connected
```
Comment (filippoweb3) [inline docs/build/build-guide.md]:
```suggestion
{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} is a blockchain protocol with two goals: providing **shared security** among all connected
```

File: docs/build/build-guide.md [hunk 2]
```diff
@@ -7,17 +7,51 @@ keywords: [build, how to build, develop, parachain, smart contract]
...
+This build guide provides four different tracks:
+
+1. [Building Parachains](#building-parachains)
+2. [Building a Pallet](#building-a-pallet)
+3. [Developing Smart Contracts](#developing-smart-contracts)
+4. [Developing a dApp](#developing-a-dapp)
+
+:::info
+
+Keep in mind that these tracks are merely suggestive, and there are many ways to utilize Polkadot,
```
Comment (filippoweb3) [inline docs/build/build-guide.md]:
```suggestion
Keep in mind that these tracks are merely suggestive, and there are many ways to utilize {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }},
```

File: docs/build/build-guide.md [hunk 3]
```diff
@@ -7,17 +7,51 @@ keywords: [build, how to build, develop, parachain, smart contract]
...
+Substrate, and their various developmental components. For more inspiration, take a look at the
+[open source projects](./build-open-source.md) featured here in the wiki!
+
+:::
+
+Before diving into any one of these tracks, it is encouraged to read about Polkadot and its networks
```
Comment (filippoweb3) [inline docs/build/build-guide.md]:
```suggestion
Before diving into any one of these tracks, it is encouraged to read about {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} and its networks
```

File: docs/build/build-guide.md [hunk 4]
```diff
@@ -44,27 +79,126 @@ But before you jump head-first into the code, you should consider the _kind_ of
 application you want to make and understand the different paradigms available to developers who want
 to build on {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }}.
 
-## What is the difference between building a parachain, a parathread, or a smart contract?
+## Building Parachains
+
+![build 2](../assets/build-2.png)
+
+As mentioned, Polkadot is the **mainnet**, but is canonically referred to as the Relay Chain. It is
```
Comment (filippoweb3) [inline docs/build/build-guide.md]:
```suggestion
{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} is canonically referred to as the Relay Chain. It is
```

File: docs/build/build-guide.md [hunk 5]
```diff
@@ -44,27 +79,126 @@ But before you jump head-first into the code, you should consider the _kind_ of
...
-help you understand them.
+![build 1](../assets/build-1.png)
+
+:::info
 
-### Parachains & Parathreads
+Throughout this document, you will encounter the term **runtime** or **STF (State Transition
+Function)**. Both refer to the same concept, as they both define how exactly a particular system,
+i.e., a blockchain, should deal with state changes both externally and internally. Both of these
+terms are used extensively in Polkadot, and by association, Substrate contexts.
```
Comment (filippoweb3) [inline docs/build/build-guide.md]:
```suggestion
terms are used extensively in {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }}, and by association, Substrate contexts.
```

**Review by filippoweb3 — CHANGES_REQUESTED** [review]
Added conditional rendering, I noticed that there are still places with "Polkadot or Kusama". As the page is used for both wikis we need to make sure cponditional rendering is present.

**Review by filippoweb3 — CHANGES_REQUESTED** [review]
Added conditional rendering, I noticed that there are still places with "Polkadot or Kusama". As the page is used for both wikis we need to make sure cponditional rendering is present.

File: docs/build/build-guide.md [hunk 6]
```diff
@@ -7,17 +7,51 @@ keywords: [build, how to build, develop, parachain, smart contract]
...
+[open source projects](./build-open-source.md) featured here in the wiki!
+
+:::
+
+Before diving into any one of these tracks, it is encouraged to read about Polkadot and its networks
+in order to gain context about the application you could make.
+
+## Development Ecosystem Overview
+
+![Path to Building a Parachain](../assets/parachain-development.png)
```
Comment (kianenigma) [inline docs/build/build-guide.md]: I find this figure a bit convoluted. Especially the lines linking things to the cumulus box is unclear to me. Can you explain further?

File: docs/build/build-guide.md [hunk 7]
```diff
@@ -44,27 +79,126 @@ But before you jump head-first into the code, you should consider the _kind_ of
...
+
+In essence, Substrate can break down a blockchain's development process by providing crucial
+building blocks of functionality, removing the need for re-engineering complex mechanisms that
+usually are involved when developing a blockchain.
+
+Substrate can be used as a basis for a parachain to connect to a relay chain like Polkadot or
+Kusama, or even as a basis to form a conventional layer one solo chain.
+
+Currently, the most streamlined way of utilizing Substrate is
+[FRAME](https://docs.substrate.io/learn/runtime-development/#frame), which conveniently allows for a
```
Comment (kianenigma) [inline docs/build/build-guide.md:176]: In the future, we should try and gradually reduce links to `docs.substrate.io` and instead rely more and more on our RUST-docs in the short term, and in the long term (paritytech.github.io) on the new doc portal.

File: docs/build/build-guide.md [hunk 8]
```diff
@@ -162,9 +274,41 @@ a smart contract.
...
+
+<TabItem value="pjs"> Promise and RxJS APIs around Polkadot and Substrate based chains via RPC
+calls. It is dynamically generated based on what the Substrate runtime provides in terms of
+metadata. Full documentation & examples available. </TabItem>
+
+<TabItem value="capi"> Capi is a framework (based on Typescript) for crafting interactions with
+Substrate chains. It consists of a development server and fluent API, which facilitates multichain
+interactions without compromising either performance or ease of use. </TabItem>
+
+<TabItem value="subxt"> Submit extrinsics (transactions) to a Substrate node via RPC using Rust.
```
Comment (kianenigma) [inline docs/build/build-guide.md:340]: 👍

File: docs/build/build-guide.md (hunk 6 above)
Comment (CrackTheCode016) [reply docs/build/build-guide.md]: The idea was to sort of show the relationships between various components, and how one *could* morph from one to another.  Pallets make up your custom runtime -> Which are used by FRAME -> Which are both built on Substrate. 

The Cumulus part was to also show that really, they are all concluded as a runtime no matter the path you take (parachain or solo chain). 

 I'll work on making t clearer, probably a top-down approach is better 👍

File: docs/build/build-guide.md (hunk 7 above)
Comment (CrackTheCode016) [reply docs/build/build-guide.md:176]: That'd be awesome.  Right now, the only equivalent to those docs in the Rust docs is this one as far as I know 😅 https://paritytech.github.io/substrate/master/frame_support/macro.construct_runtime.html

**Comment by CrackTheCode016** [comment]
@filippoweb3 addressed

File: docs/build/build-guide.md (hunk 7 above)
Comment (kianenigma) [reply docs/build/build-guide.md:176]: Once https://github.com/paritytech/substrate/pull/14137/files is done, we can use it here :)

File: docs/build/build-guide.md (hunk 6 above)
Comment (kianenigma) [reply docs/build/build-guide.md]: Any updates on this?

Comment (CrackTheCode016) [reply docs/build/build-guide.md]: I will be working on it early next week - the education repo has been taking my time :) 

I'll request your review when I complete the last set of changes 👍

Comment (CrackTheCode016) [reply docs/build/build-guide.md]: Actually, it seems a lot of that repo may be relevant here!  I'll probably tag you in some of the diagrams I'm putting together over there :) Mostly to do with `frame` and their relationship with Substrate core libraries, that kind of stuff

Comment (CrackTheCode016) [reply docs/build/build-guide.md]: Working diagram - I think it's clearer than the other one so far.  Goal was to make it super clear what does what, and how they relate to one another

Screenshot 2023-06-22 at 12 51 27 PM

Comment (CrackTheCode016) [reply docs/build/build-guide.md]: If something like this is good, then I can theme it

edit: themed.

**Comment by CrackTheCode016** [comment]
@kianenigma @DrW3RK @filippoweb3 I have polished and simplified this page and have a working diagram for the overall ecosystem development path.  

Some things to consider:

- There is a lot of info. Is there too much info? Do you think it should be more concise?
- Do the current tracks need changing? Are these what people should be looking for / to do when **first** coming to Polkadot?
- Who else could weigh in on an opinion here, as mentioned, this should and does require more than one party's input.

**Review by filippoweb3 — COMMENTED** [review]
lgtm, but I am probably not the best person to approve this PR as I still lack some knowledge about dev. Leaving final approval to Kian or Radha here.

**Review by DrW3RK — APPROVED** [review]

**Comment by DrW3RK** [comment]
This is very resourceful. Thank you, Bader! Let's merge this and share the readable Wiki doc to the relevant teams at Parity for feedback.

**Comment by CrackTheCode016** [comment]
@filippoweb3 psst, needs your approval (changes were addressed!) :)

**Review by filippoweb3 — APPROVED** [review]

File: docs/build/build-guide.md (hunk 6 above)
Comment (kianenigma) [reply docs/build/build-guide.md]: I am seeing this with a lot of delay, but the outcome is great, thanks!

**Pull request merged**